src/common/lib/types/presencemessage.ts

```diff
@@ -132,65 +87,65 @@ class PresenceMessage {
     };
   }
 
-  /**
-   * Overload toJSON() to intercept JSON.stringify()
-   * @return {*}
-   */
-  toJSON(): {
-    id?: string;
-    clientId?: string;
-    action: number;
-    data: string | Buffer | Uint8Array;
-    encoding?: string;
-    extras?: any;
-  } {
-    /* encode data to base64 if present and we're returning real JSON;
-     * although msgpack calls toJSON(), we know it is a stringify()
-     * call if it has a non-empty arguments list */
-    let data = this.data as string | Buffer | Uint8Array;
-    let encoding = this.encoding;
-    if (data && Platform.BufferUtils.isBuffer(data)) {
-      if (arguments.length > 0) {
-        /* stringify call */
-        encoding = encoding ? encoding + '/base64' : 'base64';
-        data = Platform.BufferUtils.base64Encode(data);
-      } else {
-        /* Called by msgpack. toBuffer returns a datatype understandable by
-         * that platform's msgpack implementation (Buffer in node, Uint8Array
-         * in browsers) */
-        data = Platform.BufferUtils.toBuffer(data);
-      }
-    }
-    return {
-      id: this.id,
-      clientId: this.clientId,
-      /* Convert presence action back to an int for sending to Ably */
-      action: toActionValue(this.action as string),
-      data: data,
-      encoding: encoding,
-      extras: this.extras,
-    };
+  async encode(options: CipherOptions): Promise<WirePresenceMessage> {
+    const res = Object.assign(new WirePresenceMessage(), this, {
+      action: actions.indexOf(this.action || 'present'),
+    });
+    return encode(res, options);
+  }
+
+  static fromValues(values: Properties<PresenceMessage>): PresenceMessage {
+    return Object.assign(new PresenceMessage(), values);
+  }
+
+  static fromValuesArray(values: Properties<PresenceMessage>[]): PresenceMessage[] {
+    return values.map(PresenceMessage.fromValues);
```

Comment:
I believe that in JS, it’s better to use a lambda here. The map callback has three arguments, and although it’s unlikely that we’ll add a new optional argument to `fromValues`, it could happen, and this spot would be easy to overlook.
```suggestion
    return values.map(v => PresenceMessage.fromValues(v));
```

Reply:
sure